### PR TITLE
Issue 29711 analytics matcher layer

### LINKED ISSUE #29711: Analytics: Create the Matcher Layer
### Parent Issue

[https://github.com/dotCMS/core/issues/29476](https://github.com/dotCMS/core/issues/29476)

### User Story

As part of the Analytics infrastructure; we need to create a matcher layer
The matcher layer will give us the ability to catch functionalities by catching the url, it could be an exact match, or wildcard pattern + one of more verbs (GET, POST, PUT, DELETE, etc)
In addition the clients might add matchers programatically by OSGI

### Acceptance Criteria

We should catch all vanities, urlmaps, apis and all urls we want to track as part of the analytics

### Proposed Objective

Core Features

### Proposed Priority

Priority 3 - Average

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by jdotcms**
This is an implementation for the matcher layer on analytics

This PR fixes: #29711

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/analytics/track/FilesRequestMatcher.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import com.dotmarketing.filters.CMSFilter;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.Objects;
+
+/**
+ * Matcher for pages or files
+ * @author jsanca
+ */
```
Comment (freddyDOTCMS) [inline dotCMS/src/main/java/com/dotcms/analytics/track/FilesRequestMatcher.java:15]: it is just file now

File: dotCMS/src/main/java/com/dotcms/analytics/track/AnalyticsTrackWebInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public boolean afterIntercept(final HttpServletRequest request, final HttpServletResponse response) {
+
+        if (whiteBlackList.isAllowed(request.getRequestURI())) {
+            final Optional<RequestMatcher> matcherOpt = this.anyMatcher(request, response, RequestMatcher::runAfterRequest);
+            if (matcherOpt.isPresent()) {
+
+                Logger.debug(this, () -> "afterIntercept, Matched: " + matcherOpt.get().getId() + " request: " + request.getRequestURI());
+                //fireNextStep(request, response);
```
Comment (freddyDOTCMS) [inline dotCMS/src/main/java/com/dotcms/analytics/track/AnalyticsTrackWebInterceptor.java:90]: are you going method to be executed for each RequestMatcher after and before intercept?
or this next steps is trigger the Jitsu Event?

File: dotCMS/src/main/java/com/dotcms/analytics/track/PagesAndUrlMapsRequestMatcher.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.Objects;
+
+/**
+ * Matcher for pages or files
+ * @author jsanca
+ */
+public class PagesAndUrlMapsRequestMatcher implements RequestMatcher {
```
Comment (freddyDOTCMS) [inline dotCMS/src/main/java/com/dotcms/analytics/track/PagesAndUrlMapsRequestMatcher.java:16]: Really it is just Pages, I think we need another one for URLMap

I think we can check this attribute

https://github.com/dotCMS/core/blob/bffd903770f86a8ac2da2d7e5ef07f0e158e22dc/dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/HTMLPageAssetRenderedAPIImpl.java#L448

to know if the URL was processed using URLMAP

File: dotCMS/src/main/java/com/dotcms/analytics/track/RequestMatcher.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    /**
+     * Return an id for the Matcher, by default returns the class name.
+     * @return
+     */
+    default String getId() {
+
+        return this.getClass().getName();
+    }
+}
```
Comment (freddyDOTCMS) [inline dotCMS/src/main/java/com/dotcms/analytics/track/RequestMatcher.java:126]: I suppose that all this default implementation is going to be used in cases like API and GraphQL, I think is better to keep this interface clean and move this default implementation to a sub interface or a class that implement then and later the Implementation for API and GraphQL can extends or implemented it.

Although it is just an idea

File: dotcms-integration/src/test/java/com/dotcms/analytics/track/RequestMatcherTest.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+        mockReq.setRequestURI("/api/v1/test");
+        mockReq.setRequestURL(new StringBuffer("http://localhost" + url));
+        mockReq.setServerName("http://localhost");
+        mockReq.setMethod(HttpMethod.GET);
+        final boolean result = requestMatcher.match(mockReq);
+        assertTrue(result);
+    }
+
+}
```
Comment (freddyDOTCMS) [inline dotcms-integration/src/test/java/com/dotcms/analytics/track/RequestMatcherTest.java:74]: we also need test for each Matcher implementation

File: dotCMS/src/main/java/com/dotcms/visitor/filter/logger/VisitorLogger.java [hunk 6]
```diff
@@ -118,24 +119,42 @@ public static List<Constructor<AbstractCharacter>> removeConstructor(
...
-        AbstractCharacter base = new BaseCharacter(request, response);
         try {
-            for (Constructor<AbstractCharacter> con : constructors) {
-                base = con.newInstance(base);
-            }
-
-            for (Constructor<AbstractCharacter> con : customConstructors) {
-                base = con.newInstance(base);
-            }
+            final Character base = createCharacter(request, response);
```
Comment (freddyDOTCMS) [inline dotCMS/src/main/java/com/dotcms/visitor/filter/logger/VisitorLogger.java:126]: why we need this changes here?

**Review by freddyDOTCMS — CHANGES_REQUESTED** [review]

File: dotCMS/src/main/java/com/dotcms/visitor/filter/logger/VisitorLogger.java (hunk 6 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/visitor/filter/logger/VisitorLogger.java:126]: i have encapsulate this code into a method, in order to be reused into another part

File: dotCMS/src/main/java/com/dotcms/analytics/track/AnalyticsTrackWebInterceptor.java (hunk 2 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/analytics/track/AnalyticsTrackWebInterceptor.java:90]: the request matcher has a method that tells to the Interceptor if should run before or after or both

File: dotCMS/src/main/java/com/dotcms/analytics/track/PagesAndUrlMapsRequestMatcher.java (hunk 3 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/analytics/track/PagesAndUrlMapsRequestMatcher.java:16]: The character already realize in advance if the page is an url map

File: dotCMS/src/main/java/com/dotcms/analytics/track/RequestMatcher.java (hunk 4 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/analytics/track/RequestMatcher.java:126]: Not sure if the API will use the matcher, matcher are mostly to catch requests, not api call; we suppose to create an annotation for that regard!

**Review by freddyDOTCMS — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/analytics/track/RequestMatcher.java [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+        try {
+
+            uftUri = URLDecoder.decode(requestURI, CHARSET);
+        } catch (UnsupportedEncodingException e) {
+
+            uftUri = requestURI;
+        }
+
+        return RegEX.containsCaseInsensitive(uftUri, pattern.trim());
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/analytics/track/RequestMatcher.java:86]: Is there a chance we can add some circuit breaker or something similar here? What happens if the RegEx is either invalid or too complicated and takes way too much time?

**Review by jcastro-dotcms — CHANGES_REQUESTED** [review]

**Review by dsilvam — APPROVED** [review]

**Review by victoralfaro-dotcms — APPROVED** [review]

**Review by jcastro-dotcms — APPROVED** [review]

**Pull request merged**